File: anagrama/C++/Anagrama.cpp
#include <iostream>
#include <algorithm>

//Usos
using namespace std;

//Funciones
//Esta funcion determina si dos palabras son anagramas
bool sonAnagramas(const string& palabra1, const string& palabra2) {
    // Copiamos las palabras originales para no modificarlas directamente
    string copiaPalabra1 = palabra1;
    string copiaPalabra2 = palabra2;

    // Transformamos todas las letras de las palabras a minúsculas
    transform(copiaPalabra1.begin(), copiaPalabra1.end(), copiaPalabra1.begin(), ::tolower);
    transform(copiaPalabra2.begin(), copiaPalabra2.end(), copiaPalabra2.begin(), ::tolower);

    // Ordenamos alfabéticamente las letras de ambas palabras
    sort(copiaPalabra1.begin(), copiaPalabra1.end());
    sort(copiaPalabra2.begin(), copiaPalabra2.end());

    // Comparamos las palabras ordenadas. Si son iguales, las palabras originales son anagramas.
    return copiaPalabra1 == copiaPalabra2;
}

int main() {
    // Variables para almacenar las palabras ingresadas por el usuario
    string palabra1, palabra2;

    // Solicitamos al usuario que ingrese las dos palabras
    cout << "Ingresa la primera palabra: ";
    cin >> palabra1;
    cout << "Ingresa la segunda palabra: ";
    cin >> palabra2;

    // Llamamos a la función sonAnagramas y mostramos el resultado
    bool resultado = sonAnagramas(palabra1, palabra2);
    cout << (resultado ? "Son anagramas." : "No son anagramas.") << endl;

    // Indicamos que el programa se ejecutó correctamente
    return 0;
}

File: anagrama/JavaScript/Anagrama.js
/*          ANAGRAMA
 * Escribe una función que reciba dos palabras (String) y retorne
 * verdadero o falso (Bool) según sean o no anagramas.
 * - Un Anagrama consiste en formar una palabra reordenando TODAS
 *   las letras de otra palabra inicial.
 * - NO hace falta comprobar que ambas palabras existan.
 * - Dos palabras exactamente iguales no son anagrama.
 */

//Funcion
function sonAnagramas(palabra1, palabra2) {
    // Convertimos las palabras a minúsculas y luego en arrays de caracteres
    const arrPalabra1 = palabra1.toLowerCase().split("");
    const arrPalabra2 = palabra2.toLowerCase().split("");

    // Verificamos si tienen la misma longitud
    if (arrPalabra1.length !== arrPalabra2.length) {
        return false;
    } else {
        // Ahora las ordenamos
        arrPalabra1.sort();
        arrPalabra2.sort();

        // Convertimos los arrays ordenados de nuevo a cadenas y comparamos
        return arrPalabra1.join("") === arrPalabra2.join("");
    }
}

// Solicitar al usuario que ingrese las palabras
const palabra1 = prompt("Dame la primera palabra:");
const palabra2 = prompt("Dame la segunda palabra:");

// Verificar si son anagramas
const resultado = sonAnagramas(palabra1, palabra2);
console.log(resultado);
